src/Compilers/CSharp/Test/EndToEnd/EndToEndTests.cs

```diff
@@ -514,5 +514,125 @@ string makeExpectedOutput()
+                void M2(I0 x, System.Func<string, I0> f) { }
+                // ...
+                void M2(I9 x, System.Func<string, I9> f) { }
+            */
+            for (int i = 0; i < overloads2Number; i++)
+            {
+                builder2.AppendLine($$"""void M2(I{{i}} x, System.Func<string, I{{i}}> f) { }""");
+            }
+
+            var builder3 = new StringBuilder();
+            // Local functions should be similarly fast as lambdas.
+            if (localFunctions)
+            {
+                /*
+                    M2(x, L0);
+                    static I0 L0(string arg) {
+                        arg = arg + "0";
+                        // ...
+                        arg = arg + "9";
+                        return default;
+                    }
+                    // ...
+                    M2(x, L9);
+                    static I0 L9(string arg) { ... }
+                */
+                for (int i = 0; i < lambdasNumber; i++)
+                {
+                    builder3.AppendLine($"M2(x, L{i});");
+                    builder3.AppendLine($$"""static I0 L{{i}}(string arg) {""");
+                    for (int j = 0; j < statementsNumber; j++)
+                    {
+                        builder3.AppendLine($"""arg = arg + "{j}";""");
+                    }
+                    builder3.AppendLine("return default;");
+                    builder3.AppendLine("}");
+                }
+            }
+            else
+            {
+                /*
+                    M2(x, static I0 (string arg) => {
+                        arg = arg + "0";
+                        // ...
+                        arg = arg + "9";
+                        return default;
+                    });
+                    // ...
+                    M2(x, static I0 (string arg) => {
+                        // ...
+                        return default;
+                    });
+                */
+                for (int i = 0; i < lambdasNumber; i++)
+                {
+                    builder3.AppendLine("M2(x, static I0 (string arg) => {");
+                    for (int j = 0; j < statementsNumber; j++)
+                    {
+                        builder3.AppendLine($"""arg = arg + "{j}";""");
+                    }
+                    builder3.AppendLine("return default;");
+                    builder3.AppendLine("});");
+                }
+            }
+
+            var source = $$"""
+                {{builder1}}
+                class C
+                {
+                    {{builder2}}
+                    void Main()
+                    {
+                        M1(x =>
+                        {
+                            {{builder3}}
+                        });
+                    }
+                }
+                """;
+            RunInThread(() =>
+            {
+                var comp = CreateCompilation(source, options: TestOptions.DebugDll.WithConcurrentBuild(false));
```

Comment:
>TestOptions.DebugDll.WithConcurrentBuild(false) [](http://example.com/codeflow?start=62&length=47)

Should we test `WithConcurrentBuild(true)` instead?

Reply:
We could, but I'm worried that then the number of lambda bindings would be non-deterministic as two threads could race to bind the same lambda so the counter would be incremented twice in such rare case, making the test flaky.